drivers/sensor/nxp/nxp_mcx_temp/temp_mcx.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_mcx_temperature
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/devicetree.h>
+
+LOG_MODULE_REGISTER(temp_mcx, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Four ADC samples: first two are dropped, then VBE1 and VBE8 */
+#define TEMP_MCX_ADC_SAMPLES 1
+
+struct temp_mcx_config {
+	const struct device *adc;
+	struct adc_sequence adc_seq;
+	struct adc_channel_cfg ch_cfg;
+};
+
+struct temp_mcx_data {
+	float temperature;
+	uint16_t buffer[TEMP_MCX_ADC_SAMPLES];
+};
+
+static int temp_mcx_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct temp_mcx_config *config = dev->config;
+	struct temp_mcx_data *data = dev->data;
+	int err;
+	uint16_t vbe1 = 0U;
+	uint16_t vbe8 = 0U;
+	uint32_t conv_result_shift = 0U;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
+	}
+
+#if defined(FSL_FEATURE_LPADC_TEMP_SENS_BUFFER_SIZE) &&                                            \
+	(FSL_FEATURE_LPADC_TEMP_SENS_BUFFER_SIZE == 4U)
+	/* The firsttwo results is useless. */
+	adc_read(config->adc, &config->adc_seq);
+	adc_read(config->adc, &config->adc_seq);
+#endif /* FSL_FEATURE_LPADC_TEMP_SENS_BUFFER_SIZE */
+
+	err = adc_read(config->adc, &config->adc_seq);
+	if (err) {
+		LOG_ERR("Failed to read ADC channels (err %d)", err);
+		return err;
+	}
+	vbe1 = data->buffer[0] >> conv_result_shift;
+
+	err = adc_read(config->adc, &config->adc_seq);
+	if (err) {
+		LOG_ERR("Failed to read ADC channels (err %d)", err);
+		return err;
+	}
+	vbe8 = data->buffer[0] >> conv_result_shift;
+
+	/* Calculate temperature using exact MCX formula */
+	data->temperature =
+		FSL_FEATURE_LPADC_TEMP_PARAMETER_A *
+			(FSL_FEATURE_LPADC_TEMP_PARAMETER_ALPHA * ((float)vbe8 - (float)vbe1) /
+			 ((float)vbe8 +
+			  FSL_FEATURE_LPADC_TEMP_PARAMETER_ALPHA * ((float)vbe8 - (float)vbe1))) -
+		FSL_FEATURE_LPADC_TEMP_PARAMETER_B;
+
+	LOG_DBG("VBE1=%d VBE8=%d Temp=%.3f", vbe1, vbe8, (double)data->temperature);
+	return 0;
+}
+
+static int temp_mcx_channel_get(const struct device *dev, enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct temp_mcx_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
```

Comment:
not sure if this is correct or matters, but maybe it's supposed to return -EINVAL here, since it's not that it's not supported in the software but that it is an invalid channel type for this hardware. @MaureenHelm ?